### DISCUSSION TITLE
Collaboration

### DISCUSSION THREAD

**Discussion by FrancescoSaverioZuppichini** [frgfm/Holocron#143] [Ideas]
Hello there, 

I am the author of [glasses](https://github.com/FrancescoSaverioZuppichini/glasses) what if we join forces?  

May the force be with you!

Francesco

**Comment by frgfm** [comment]
Hi there @FrancescoSaverioZuppichini :wave: 

Thanks for the suggestion! I'm always happy working with new people, always bring opportunities to improve one's methods / ways of designing tools. 

However, what do you have in mind specifically? Do you mean for training procedures? For implementations? or some one-shot collaboration on a demo or something?

**Reply by frgfm** [reply]
Figured it would be more appropriate in discussions rather than issues :) @FrancescoSaverioZuppichini

**Reply by FrancescoSaverioZuppichini** [reply]
Hi @frgfm , first of all, sorry but I totally stole your obi-wan git for my personal Readme :) 

What's is your end goal? Mine is to create a library of well-written reusable components for computer vision with common APIs and in the process learn as much as I can. 

Honestly, my library needs an afternoon of polishing but I think it is in a good state. So, yeah, I was looking around for guys like you to join forces and make something great together. 

If you look at the current most used CV libraries like [this](https://github.com/rwightman/pytorch-image-models) they kinda suck. They are badly written and every model has a different API. I solve it by following a general structure that makes sense, composition over inheritance etc. Then, since pre-trained weights are linked to the `state_dict` of a model and if you want to directly load them you have to have the same "design". I traced the operations and converted the weights to make them compatible with my implementations, so far I've never seen other people doing it.

Thank you again :) 

Take care!

**Reply by frgfm** [reply]
How dare you? :hushed:
You don't need GIF mastery to find other relevant GIFs for a readme haha

About goals:
- happy to help with the polishing at some point :+1: 
- not sure I can see how (if that's a good idea) to merge the repos, but helping each other could be a good way to start!
- Careful now, if you say Ross Wightman's projects suck, we're gonna have a problem haha. Not all projects are for entry level users, Ross focuses on developers, he trains so many models beyond SOTA for the part of the community that knows how to use them. On the other side, you can choose to make more accessible projects (like I try to do), but I would argue that Ross has made a huge contribution to the DL community :smile:  (by the way, he does also remap existing weights before being able to train new ones)

**Reply by FrancescoSaverioZuppichini** [reply]
Hey @frgfm , how are you doing?

- Cool! Are you up for a call, you seem like a super interesting guy and I would love to connect!
- It would be amazing if you can give me feedbacks ;) 
- I don't fully agree (that is an interesting discussion). Like any good lightsaber, they are useless if you cannot use them! So Ross is a hero, no doubt, but his library is not developers' focused because there is no "sharability" in its code. Most of the time is copy & pasted from the original implementations. No common APIs = hard to use/share. This is why it sucks, research should be made with the focus of sharing and letting other people use it, not self-showing. This is exactly what I want to improve, I want people to have an easier time with CV, why should two models have two different APIs if they solve the same task? 

What do you think? 

Thank you :)

**Reply by frgfm** [reply]
@FrancescoSaverioZuppichini I've been away for a few weeks, my bad!

- Next week or the following one, happy to do so!
- I'll take a look next week :+1: 
- I agree but just like with the force, there is a need for some balance :sweat_smile: Flexible enough for researchers to iterate/finetune, accessible enough for padawans. I try to come up with a common API on the models I implement, but leaving enough possibilities to avoid forcing developers to reimplement everything for that to work
